mdal/mdal_data_model.cpp

```diff
@@ -8,20 +8,91 @@
 #include <algorithm>
 #include "mdal_utils.hpp"
 
-bool MDAL::Dataset::isActive( size_t faceIndex )
+#define NODATA std::numeric_limits<double>::quiet_NaN()
+
+MDAL::Dataset::~Dataset() = default;
+
+size_t MDAL::Dataset::valuesCount() const
 {
   assert( parent );
-  if ( parent->isOnVertices )
+  assert( parent->parent );
+  if ( parent->isOnVertices() )
   {
-    if ( active.size() > faceIndex )
-      return active[faceIndex];
-    else
-      return false;
+    return parent->parent->verticesCount();
+  }
+  else
+  {
+    return parent->parent->facesCount();
+  }
+}
+
+MDAL::MemoryDataset::~MemoryDataset() = default;
+
+size_t MDAL::MemoryDataset::activeData( size_t indexStart, size_t count, char *buffer )
+{
+  assert( parent );
+  if ( parent->isOnVertices() )
+  {
+    assert( active.size() > indexStart ); //checked in C API interface
+    assert( active.size() >= indexStart + count ); //checked in C API interface
+    char *src = active.data() + indexStart;
+    memcpy( buffer, src, count );
   }
   else
   {
+    memset( buffer, true, count );
     return true;
   }
+
+  return count;
+}
+
+size_t MDAL::MemoryDataset::scalarData( size_t indexStart, size_t count, double *buffer )
+{
+  assert( parent ); //checked in C API interface
+  assert( parent->isScalar() ); //checked in C API interface
+  assert( values.size() > indexStart ); //checked in C API interface
+  assert( values.size() >= indexStart + count ); //checked in C API interface
+
+  for ( size_t i = 0; i < count; ++i )
+  {
+    const MDAL::Value value = values[ indexStart + i ];
+    if ( value.noData )
+    {
+      buffer[i] = NODATA;
+    }
+    else
+    {
+      buffer[i] = value.x;
```

Comment:
Maybe for scalar datasets we do not need to keep x+y to save some memory?

Reply:
it should be reworked a lot, but I would keep it for separate PR/task since it is not related strictly to the lazy loading